src/app/rpmostree-polkit-agent.c

```diff
@@ -0,0 +1,257 @@
+ * Copyright (C) 2012 Matthias Klumpp <matthias@tenstral.net>
+ *
+ * Licensed under the GNU Lesser General Public License Version 2.1
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with systemd; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/* snipped from PackageKit/systemd */
+
+#include <stdio.h>
+#include <sys/types.h>
+#include <sys/prctl.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <string.h>
+#include <dirent.h>
+#include <signal.h>
+#include <fcntl.h>
+#include <errno.h>
+#include <inttypes.h>
+#include <sys/poll.h>
+#include <sys/wait.h>
+#include <glib.h>
+
+#include "rpmostree-polkit-agent.h"
+
+#define memzero(x,l) (memset((x), 0, (l)))
+#define zero(x) (memzero(&(x), sizeof(x)))
+
+#define POLKIT_TTY_AGENT_BINARY_PATH "/usr/bin/pkttyagent"
+
+static pid_t agent_pid = 0;
+
+static int
+fork_agent (pid_t *pid, const char *path, ...)
+{
+  pid_t parent_pid, n_agent_pid;
+  int fd;
+  gboolean stdout_is_tty, stderr_is_tty;
+  unsigned n, i;
+  va_list ap;
+  char **l;
+
+  g_return_val_if_fail (pid != 0, 0);
+  g_assert (path);
+
+  parent_pid = getpid ();
+
+  /* Spawns a temporary TTY agent, making sure it goes away when
+   * we go away */
+
+  n_agent_pid = fork ();
+  if (n_agent_pid < 0)
+    return -errno;
+
+  if (n_agent_pid != 0)
+    {
+      *pid = n_agent_pid;
+      return 0;
+    }
+
+  /* In the child:
+   *
+   * Make sure the agent goes away when the parent dies */
+  if (prctl (PR_SET_PDEATHSIG, SIGTERM) < 0)
+    _exit (EXIT_FAILURE);
+
+  /* Check whether our parent died before we were able
+   * to set the death signal */
+  if (getppid () != parent_pid)
+    _exit (EXIT_SUCCESS);
+
+  /* TODO: it might be more clean to close all FDs so we don't leak them to the agent */
+
+  stdout_is_tty = isatty (STDOUT_FILENO);
+  stderr_is_tty = isatty (STDERR_FILENO);
+
+  if (!stdout_is_tty || !stderr_is_tty)
+    {
+      /* Detach from stdout/stderr. and reopen
+       * /dev/tty for them. This is important to
+       * ensure that when systemctl is started via
+       * popen() or a similar call that expects to
+       * read EOF we actually do generate EOF and
+       * not delay this indefinitely by because we
+       * keep an unused copy of stdin around. */
+      fd = open("/dev/tty", O_WRONLY);
+      if (fd < 0)
+        {
+          g_error ("Failed to open /dev/tty: %m");
```

Comment:
Use `errx`, we shouldn't dump core here, and also allows dropping the `_exit()`.